File: src/main/scala/com/martomate/tripaint/control/TriPaintController.scala
package com.martomate.tripaint.control

import com.martomate.tripaint.model.coords.GridCoords
import com.martomate.tripaint.model.effects.*
import com.martomate.tripaint.model.image.{GridCell, ImageStorage}
import com.martomate.tripaint.model.{Color, TriPaintModel}
import com.martomate.tripaint.view.gui.UIAction
import com.martomate.tripaint.view.{TriPaintView, TriPaintViewFactory, TriPaintViewListener}

class TriPaintController(val model: TriPaintModel, viewFactory: TriPaintViewFactory)
    extends TriPaintViewListener {
  val view: TriPaintView = viewFactory.createView(this, model)

  override def perform(action: UIAction): Unit = action match
    case UIAction.New =>
      for
        (x, y) <- view.askForWhereToPutImage()
        backgroundColor = Color.fromFXColor(view.backgroundColor)
        coords = GridCoords(x, y)
      do Actions.createNewImage(model.imageGrid, backgroundColor, coords)

    case UIAction.Open =>
      for
        file <- view.askForFileToOpen()
        fileOpenSettings <- view.askForFileOpenSettings(file, model.imageGrid.imageSize, 1, 1)
        (x, y) <- view.askForWhereToPutImage()
        coords = GridCoords(x, y)
      do Actions.openImage(model, file, fileOpenSettings, coords)

    case UIAction.OpenHexagon =>
      for
        file <- view.askForFileToOpen()
        fileOpenSettings <- view.askForFileOpenSettings(file, model.imageGrid.imageSize, 6, 1)
        (x, y) <- view.askForWhereToPutImage()
        coords = GridCoords(x, y)
      do Actions.openHexagon(model, file, fileOpenSettings, coords)

    case UIAction.Save =>
      Actions.save(
        model.imagePool,
        model.imageGrid.selectedImages.filter(_.changed),
        model.fileSystem
      )(view.askForSaveFile, view.askForFileSaveSettings, view)

    case UIAction.SaveAs =>
      model.imageGrid.selectedImages.foreach(im =>
        Actions.saveAs(model.imagePool, im, model.fileSystem)(
          view.askForSaveFile,
          view.askForFileSaveSettings,
          view
        )
      )

    case UIAction.Exit => if do_exit() then view.close()

    case UIAction.Undo => model.imageGrid.undo()

    case UIAction.Redo => model.imageGrid.redo()

    case UIAction.Blur =>
      for radius <- view.askForBlurRadius()
      do Actions.applyEffect(model, new BlurEffect(radius))

    case UIAction.MotionBlur =>
      for radius <- view.askForMotionBlurRadius()
      do Actions.applyEffect(model, new MotionBlurEffect(radius))

    case UIAction.RandomNoise =>
      for (lo, hi) <- view.askForRandomNoiseColors()
      do Actions.applyEffect(model, new RandomNoiseEffect(lo, hi))

    case UIAction.Scramble => Actions.applyEffect(model, ScrambleEffect)

    case _ =>

  override def requestExit(): Boolean = do_exit()

  override def requestImageRemoval(image: GridCell): Unit =
    var abortRemoval = false
    if image.changed then
      view.askSaveBeforeClosing(Seq(image)) match
        case Some(shouldSave) =>
          if (
            shouldSave && !Actions.save(model.imagePool, Seq(image), model.fileSystem)(
              view.askForSaveFile,
              view.askForFileSaveSettings,
              view
            )
          ) abortRemoval = true
        case None => abortRemoval = true

    if (!abortRemoval) model.imageGrid -= image.coords

  private def do_exit(): Boolean =
    model.imageGrid.changedImages match
      case Seq() => true
      case images =>
        view.askSaveBeforeClosing(images) match
          case Some(shouldSave) =>
            if shouldSave then
              Actions.save(model.imagePool, images, model.fileSystem)(
                view.askForSaveFile,
                view.askForFileSaveSettings,
                view
              )
            else true
          case None => false
}

File: app/src/main/kotlin/tripaint/app/TriPaintController.kt
package tripaint.app

import javafx.application.Platform
import javafx.stage.Stage
import tripaint.coords.GridCoords
import tripaint.effects.*
import tripaint.grid.GridCell
import tripaint.grid.ImageGrid
import tripaint.image.ImagePool
import tripaint.view.TriPaintView
import tripaint.view.TriPaintViewListener
import tripaint.view.gui.UIAction

class TriPaintController(stage: Stage, private val fileSystem: FileSystem) : TriPaintViewListener {
    private val imagePool: ImagePool = ImagePool()
    private val imageGrid: ImageGrid = ImageGrid(-1)

    private val view: TriPaintView = MainStage(this, fileSystem, imagePool, imageGrid, stage)

    init {
        Platform.runLater {
            imageGrid.setImageSizeIfEmpty(view.askForImageSize() ?: 32)
        }
    }

    override fun perform(action: UIAction) {
        when (action) {
            UIAction.New -> {
                view.askForWhereToPutImage()?.let { (x, y) ->
                    val backgroundColor = view.backgroundColor()
                    val coords = GridCoords.from(x, y)
                    Actions.createNewImage(imageGrid, backgroundColor, coords)
                }
            }

            UIAction.Open -> {
                view.askForFileToOpen()?.let { file ->
                    view.askForFileOpenSettings(file, imageGrid.imageSize, 1, 1)?.let { fileOpenSettings ->
                        view.askForWhereToPutImage()?.let { (x, y) ->
                            val coords = GridCoords.from(x, y)
                            Actions.openImage(fileSystem, imagePool, imageGrid, file, fileOpenSettings, coords)
                        }
                    }
                }
            }

            UIAction.OpenHexagon -> {
                view.askForFileToOpen()?.let { file ->
                    view.askForFileOpenSettings(file, imageGrid.imageSize, 6, 1)?.let { fileOpenSettings ->
                        view.askForWhereToPutImage()?.let { (x, y) ->
                            val coords = GridCoords.from(x, y)
                            Actions.openHexagon(fileSystem, imagePool, imageGrid, file, fileOpenSettings, coords)
                        }
                    }
                }
            }

            UIAction.Save -> {
                Actions.save(
                    imageGrid,
                    imagePool,
                    imageGrid.selectedImages().filter { it.changed },
                    fileSystem,
                    { image -> view.askForSaveFile(image) },
                    { file, image -> view.askForFileSaveSettings(file, image) },
                    view,
                )
            }

            UIAction.SaveAs -> {
                for (im in imageGrid.selectedImages()) {
                    Actions.saveAs(
                        imageGrid,
                        imagePool,
                        im,
                        fileSystem,
                        { image -> view.askForSaveFile(image) },
                        { file, image -> view.askForFileSaveSettings(file, image) },
                        view
                    )
                }
            }

            UIAction.Exit -> {
                if (doExit()) {
                    view.close()
                }
            }

            UIAction.Undo -> {
                imageGrid.undo()
            }

            UIAction.Redo -> {
                imageGrid.redo()
            }

            UIAction.Blur -> {
                view.askForBlurRadius()?.let { radius ->
                    Actions.applyEffect(imageGrid, BlurEffect(radius))
                }
            }

            UIAction.MotionBlur -> {
                view.askForMotionBlurRadius()?.let { radius ->
                    Actions.applyEffect(imageGrid, MotionBlurEffect(radius))
                }
            }

            UIAction.RandomNoise -> {
                view.askForRandomNoiseColors()?.let { (lo, hi) ->
                    Actions.applyEffect(imageGrid, RandomNoiseEffect(lo, hi))
                }
            }

            UIAction.Scramble -> {
                Actions.applyEffect(imageGrid, ScrambleEffect)
            }

            UIAction.Cell -> {
                Actions.applyEffect(imageGrid, CellEffect())
            }

            else -> {}
        }
    }

    override fun requestExit(): Boolean = doExit()

    override fun requestImageRemoval(image: GridCell) {
        var abortRemoval = false
        if (image.changed) {
            val shouldSave = view.askSaveBeforeClosing(listOf(image))
            if (shouldSave != null) {
                if (shouldSave && !Actions.save(
                        imageGrid,
                        imagePool,
                        listOf(image),
                        fileSystem,
                        { view.askForSaveFile(it) },
                        { file, im -> view.askForFileSaveSettings(file, im) },
                        view,
                    )
                ) {
                    abortRemoval = true
                }
            } else {
                abortRemoval = true
            }
        }

        if (!abortRemoval) {
            imageGrid.remove(image.coords)
        }
    }

    private fun doExit(): Boolean {
        val images = imageGrid.changedImages()
        if (images.isEmpty()) return true

        val shouldSave = view.askSaveBeforeClosing(images)
        return if (shouldSave != null) {
            if (shouldSave) {
                Actions.save(
                    imageGrid, imagePool, images, fileSystem,
                    { image -> view.askForSaveFile(image) },
                    { file, image -> view.askForFileSaveSettings(file, image) },
                    view
                )
            } else true
        } else false
    }
}
